### PR TITLE
feat: Add support for `CONDA_OVERRIDE_CUDA`

### PR DESCRIPTION
**Pull request by SobhanMP**
### Description

Hi, I know this is terrible, not elegant, etc, but I do think that having support for something in the spirit of CONDA_OVERRIDE_CUDA is necessary on academic HPC environments. Reasons include 
- only having internet access on the login node with no gpu
- launching jobs with submitit on login nodes with no gpu.

I'm not really used to writing rust, so would be greatful for some guidance.

Thanks

### CODE REVIEW

File: crates/rattler_virtual_packages/src/cuda.rs
Comment (baszalmstra) [inline crates/rattler_virtual_packages/src/cuda.rs:33]: This function should be pure functional without environmental side-effects.

Instead I propose we add some sort of `from_env` function to the appropriate virtual packages that allow users to choose whether they want this behavior. Or a `or_override_with_env_var` or something along those lines.

We should also debate the environment variable name. I believe conda has a different name? Im also not sure whether we should hardcode that in rattler itself.

**Comment by wolfv** [comment]
I think this would also be good for all the other virtual packages as a way to hot-fix certain scenarios that we haven't covered yet (#815 for example).

**Comment by baszalmstra** [comment]
My idea would be to add functions to the virtual package implementations. Lets take CUDA as an example, we would add these functions to the `Cuda` impl:

```rust
impl Cuda {
    pub fn from_env_var_name(env_var_name: &str) -> Option<Self> {
		let version_str = env::var(env_var_name)?;
		// Extract versions
		...
    }
    
    pub fn from_default_env_var() -> Option<Self> {
    	Self::from_env_var_name("CONDA_OVERRIDE_CUDA")
    }
    
    ...
}
```

You can then write:

```
let cuda = Cuda::from_default_env_var().or_else(Cuda::current);
```

To get the CUDA version of the system based on env var, or if not present by detecting it.

Similarly a function `VirtualPackage::from_default_env_vars` could be added. 

I think this allows:

* Users of the library to decide whether they want to allow environment overrides.
* Users can specify custom environment variables if they want to.
* Users can use the default environment variable names.

Which I think hits the sweet spot.

And indeed as @wolfv said, it would be nice to add these for the other virtual packages as well.

**Comment by SobhanMP** [comment]
oh, yeah that looks great!

**Comment by SobhanMP** [comment]
Does this make sense as a rough draft?

File: crates/rattler_virtual_packages/src/lib.rs
Comment (baszalmstra) [inline crates/rattler_virtual_packages/src/lib.rs:70]: I dont think relying on type names is good here. I prefer just reimplementing this per type instead of using a blanket implementation

Comment (baszalmstra) [reply crates/rattler_virtual_packages/src/lib.rs:70]: The name should also be [`CONDA_OVERRIDE_CUDA`](https://github.com/conda/conda/blob/77aa3d337963fdbe5c817304a2052ae0125fd326/docs/source/user-guide/tasks/manage-virtual.rst#overriding-detected-packages)

File: crates/rattler_virtual_packages/src/lib.rs
Comment (baszalmstra) [inline crates/rattler_virtual_packages/src/lib.rs:266]: Conda only allows overriding the glibc family version it seems:

https://github.com/conda/conda/blob/77aa3d337963fdbe5c817304a2052ae0125fd326/docs/source/user-guide/tasks/manage-virtual.rst#overriding-detected-packages

**Review by baszalmstra — COMMENTED** [review]
Yeah that looks a lot better! I left a few comments.

File: crates/rattler_virtual_packages/src/lib.rs
Comment (SobhanMP) [reply crates/rattler_virtual_packages/src/lib.rs:266]: Does rattler need to adhere to that?

File: crates/rattler_virtual_packages/src/lib.rs
Comment (SobhanMP) [reply crates/rattler_virtual_packages/src/lib.rs:70]: nice catch. The blanket implementation is opt in, is it really a problem?

**Comment by SobhanMP** [comment]
I think there is one issue, finding the environment variable and it being empty is different than the environment variable not being present, should the type be `Option<Option<T>>`? first one option showing that the env was found the second one that the env was none?

**Comment by SobhanMP** [comment]
I think everything should have been handled gracefully.

**Comment by SobhanMP** [comment]
Do we want to override the glibc runtime? like `CONDA_OVERRIDE_GLIBC="2@musl"`, same for `Linux` and `Archspec`. We could also do unix and win but all of these are not overridable by conda.

**Review by baszalmstra — APPROVED** [review]
This looks good! Ill take a closer look at the CI failure tomorrow!

**Comment by SobhanMP** [comment]
This should solve some of the failure, pre-commit is not ran by default

**Comment by SobhanMP** [comment]
I don't think pyrattler has any implication for pixi, so I'd rather this be merged w/o py-rattler support. My understanding is that  in `py-rattler/src/virtual_package.rs` the PyVirtualPackage needs to expose some more code, the problem is that this either requires either
1- A special classs for virtual packages that can be overriden (there is already one for arch), 
2- that all virtual pacakges have override, or 
3- using `Cuda::from_default_env_var.unwrap_or(Cuda::current().into()).unwrap()` in the `VirtualPackage::current` or `try_detect_virtual_packages` instead of `Cuda::current()`

**Comment by SobhanMP** [comment]
@baszalmstra It's very inconvenient that the checks require maintainer auth.

**Comment by baszalmstra** [comment]
Yes, but once your first PR lands that restriction is lifted 👍

**Comment by baszalmstra** [comment]
I think it would be good to also expose this functionality to the python bindings, but we can do that in another PR.

Thanks for your first contribution @SobhanMP !

**Comment by SobhanMP** [comment]
@baszalmstra I think this first requires figuring out https://github.com/conda/rattler/pull/818#issuecomment-2297850077 . I think option 2 makes the most sense.

**Comment by SobhanMP** [comment]
The next thing is where this should be used. Should  `try_detect_virtual_packages` be patched? Should it take some structure to rename overrides?

**Pull request merged**
